java/compaction/compaction-job-execution/src/main/java/sleeper/compaction/job/execution/CompactionJobMessageHandler.java

```diff
@@ -0,0 +1,138 @@
+
+import sleeper.compaction.job.CompactionJob;
+import sleeper.configuration.properties.instance.InstanceProperties;
+
+import java.io.IOException;
+import java.time.Duration;
+import java.time.Instant;
+import java.util.Optional;
+import java.util.function.Supplier;
+
+import static sleeper.configuration.properties.instance.CompactionProperty.COMPACTION_TASK_MAX_CONSECUTIVE_FAILURES;
+import static sleeper.configuration.properties.instance.CompactionProperty.COMPACTION_TASK_MAX_TIME_IN_SECONDS;
+
+public class CompactionJobMessageHandler {
+    private static final Logger LOGGER = LoggerFactory.getLogger(CompactionJobMessageHandler.class);
+
+    private final Supplier<Instant> timeSupplier;
+    private final int maxConsecutiveFailures;
+    private final int maxTimeInSeconds;
+    private final MessageReceiver messageReceiver;
+    private final MessageConsumer messageConsumer;
+    private final FailedJobHandler failedJobHandler;
+
+    public CompactionJobMessageHandler(InstanceProperties instanceProperties, Supplier<Instant> timeSupplier,
+            MessageReceiver messageReceiver, MessageConsumer messageConsumer, FailedJobHandler failedJobHandler) {
+        maxTimeInSeconds = instanceProperties.getInt(COMPACTION_TASK_MAX_TIME_IN_SECONDS);
+        maxConsecutiveFailures = instanceProperties.getInt(COMPACTION_TASK_MAX_CONSECUTIVE_FAILURES);
+        this.timeSupplier = timeSupplier;
+        this.messageReceiver = messageReceiver;
+        this.messageConsumer = messageConsumer;
+        this.failedJobHandler = failedJobHandler;
+    }
+
+    public Result runAt(Instant startTime) throws InterruptedException, IOException {
+        Instant maxTime = startTime.plus(Duration.ofSeconds(maxTimeInSeconds));
+        int numConsecutiveFailures = 0;
+        Instant currentTime = startTime;
+        long totalNumberOfMessagesProcessed = 0;
+        while (numConsecutiveFailures < maxConsecutiveFailures && currentTime.isBefore(maxTime)) {
+            Optional<JobAndMessage> jobAndMessageOpt = messageReceiver.receiveMessage();
+            if (!jobAndMessageOpt.isPresent()) {
+                numConsecutiveFailures++;
+            } else {
+                JobAndMessage jobAndMessage = jobAndMessageOpt.get();
+                LOGGER.info("CompactionJob is: {}", jobAndMessage.job);
+                try {
+                    messageConsumer.consume(jobAndMessage);
+                    totalNumberOfMessagesProcessed++;
+                    numConsecutiveFailures = 0;
+                } catch (Exception e) {
+                    LOGGER.error("Failed processing compaction job, putting job back on queue", e);
+                    numConsecutiveFailures++;
+                    failedJobHandler.onFailure(jobAndMessage);
+                }
+            }
+            currentTime = timeSupplier.get();
+        }
+        return new Result(totalNumberOfMessagesProcessed, currentTime.isAfter(maxTime), numConsecutiveFailures >= maxConsecutiveFailures);
```

Comment:
Rather than exposing the parameters in the Result class outside of here, we could just move the log messages inside this class.

It seems like the test assertions on the Result object are effectively duplicated by the setup and assertions on the data used to fake the MessageReceiver, MessageConsumer, FailedJobHandler, time supplier? It seems like those setup and assertions on the injected behaviour should be more accurate than examining the state returned here anyway?